---
title: "<i class='fas fa-flask' style='margin-left:4px;margin-right:4px;'></i> Lab"
output:
  bookdown::html_document2:
    toc: true
    toc_float: true
    toc_depth: 4
    number_sections: true
    theme: flatly
    highlight: tango
    df_print: default
    code_folding: "none"
    self_contained: false
    keep_md: false
    encoding: 'UTF-8'
    css: "assets/lab.css"
---

```{r,child="assets/header-lab.Rmd"}
```

This page contains links to different tutorials that are used in this course. The tutorials are well documented and should be easy to follow.

<div class="instruction">
Input code blocks are displayed like shown below. The code language is displayed above the block. Shell scripts (**SH**) are to be executed in the linux terminal such as bash. **R** scripts are to be run in R either through the terminal, RGui or RStudio.

```{sh,eval=FALSE,block.title=TRUE}
command
```

<i class="fas fa-exclamation-circle"></i> Note &nbsp; <i class="fas fa-lightbulb"></i> Tip &nbsp; <i class="fas fa-comments"></i> Discuss &nbsp; <i class="fas fa-clipboard-list"></i> Task
</div>

<br>

This topic covers retrieving data needed for all the exercises in the entire course can be found in the following link. 

[<i class="fas fa-link"></i> <span class="large">Downloading data</span>](lab_download.html)

Some pre-course R warm-up below, if you fancy that:

[<i class="fas fa-link"></i> <span class="large">R warmup</span>](https://immuntech.github.io/Bioinformatics_Workshop_ImmTech/Data/2019-10-15/R_exercise_Lokesh.html)

# Introduction

Most of the analyses is carried out in R and it will be useful to learn some basics in R that are very much necessary for this course. Even if you have learnt these things already, it is good to freshen up your memory. 

[<i class="fas fa-link"></i> <span class="large">Introduction to R</span>](lab_r.html)  


# Main lab

## Data

In most of the exercises, we will use RNA-seq data (Illumina short reads) from the human A431 cell line. It is an epidermoid carcinoma cell line which is often used to study cancer and the cell cycle, and as a sort of positive control of epidermal growth factor receptor (EGFR) expression. A431 cells express very high levels of EGFR, in contrast to normal human fibroblasts.

The A431 cells were treated with gefinitib, which is an EGFR inhibitor and is used (under the trade name Iressa) as a drug to treat cancers with mutated and overactive EGFR. In the experiment, RNA was extracted at four time points: before the gefinitib treatment (t=0), and two, six and twenty-four hours after treatment (t=2, t=6, t=24, respectively), and sequenced using an Illumina HiSeq instrument in triplicates (thus there are 3x4=12 samples).

This data is part of the transcriptomics course that is also given by NBIS. We will use some of the `counts table` that was generated in the course after different transformations like manual filtering for low counts, VST and DESeq2. You don't have to know what these exactly mean to do these exercises :) These are basically the same data at different stages of the transcriptomics analysis.    

## ggplot basics 1

Below is the link for the basic plotting exercise. there will be some basic plotting exercises first using R base graphics and comparing them to the grid graphics using `ggplot2`. Followed by practicing the first basics in ggplot: geoms, colors and aesthetics.

[<i class="fas fa-link"></i> <span class="large">Geoms, colors and aesthetics</span>](lab_base_grid.html)

## ggplot basics 2

Below is the link for the second exercise where you will look into facets, barplots and errorbars.

[<i class="fas fa-link"></i> <span class="large">Facets, barplots and errorbars</span>](lab_gg_2.html)

## ggplot basics 3

Below is the link for the third exercise where you will look into facets, barplots and errorbars.

[<i class="fas fa-link"></i> <span class="large">Axes, labels, legends and themes</span>](lab_gg_3.html)

## PCA and Heatmaps

Here, we look into building PCA plots with `ggplot`. We will also look into making heatmaps with both `pheatmap` and `ggplot`.

[<i class="fas fa-link"></i> <span class="large">PCA and Heatmap</span>](lab_pca_hmap.html)

**End of document**
