---
title: "Downloads"
subtitle: "Workshop on ggplot"
author: "`r paste0('<b>Lokesh Mano</b> â€¢ ',format(Sys.time(), '%d-%b-%Y'))`"
output:
  bookdown::html_document2:
          toc: true
          toc_float: true
          toc_depth: 4
          number_sections: true
          theme: flatly
          highlight: tango
          df_print: default
          code_folding: "none"
          self_contained: false
          keep_md: false
          encoding: 'UTF-8'
          css: "assets/lab.css"
---

```{r,child="assets/header-lab.Rmd"}
```

```{r,include=FALSE}
# data handling
library(dplyr)
#library(tidyr)
#library(stringr)

# plotting
library(ggplot2)

#library(biomaRt) # annotation
#library(DESeq2) # rna-seq
#library(edgeR) # rna-seq
```

# Main exercises

The following data will be used in the used for the exercises in this course. The data comes from a mock RNA sequencing data with 12 samples that are from cell-lines. As in any normal RNAseq analysis the counts of genes were normalized using different methods such as `VST` and `CPM` after filtering out the genes that were of very low expression. We would use these data for visulaizations. 

**Download all of these files into your current working directory**

You can download all the files together [here](data/data.zip).
Make a directory called `data` and `unzip` in that directory!

If `ggplot_geneco_course` is the current working directory, the directory tree should look like this:

* <span style="color:blue">ggplot_geneco_course</span>
  + <span style="color:blue">data</span>
    + arch_newick.txt
    + counts_deseq2.txt
    + counts_filtered.txt
    + counts_raw.txt
    + counts_vst.txt
    + human_biomaRt_annotation.csv
    + metadata_raw.csv
    + Time_t24_vs_t0.txt
    + Time_t2_vs_t0.txt
    + Time_t6_vs_t0.txt
    + tree_env.tsv
    + tree_hmap.tsv

You can find the information related to the important files below:

## Counts table

* Table with gene counts after filtering: [Filtered Counts](data/counts_filtered.txt)
* Table with gene counts normalized with VST: [VST counts](data/counts_vst.txt)
* Table with gene counts normalized with DESeq2: [DESeq2 counts](data/counts_deseq2.txt)

## Metadata

* Metadata of the samples: [Sample Metadata](data/metadata_raw.csv)
* Metadata of the genes with their functions: [Gene Annotation](data/human_biomaRt_annotation.csv)

## DE genes

Below are the lists of differentially expressed genes between different time points

* [Time point 2 vs Time point 0](data/Time_t2_vs_t0.txt)
* [Time point 6 vs Time point 0](data/Time_t6_vs_t0.txt)
* [Time point 24 vs Time point 0](data/Time_t24_vs_t0.txt)

# Optional exercises

## Phylogenetic trees

The data required for the phylogenetic trees are already part of the main `data.zip` file, as shown in the directory tree before. The files needed for this exercise are described below:

* A phylogentic tree based on a group of archaeal genomes in newick format: [Archaea tree](data/arch_newick.txt)
* Some metadata info regarding the environments that these archaea could be found: [Environment info](data/tree_env.tsv)
* An expression matrix to show heatmap and phylogenetic tree together: [for heatmap](data/tree_hmap.tsv)

## Map-data with ggmap

The data required for this part of the exercise can be downloaded from [here](data/ggmap_data/ggmap_data.zip)

* A small data frame of points from the [Sisquoc River](data/ggmap_data/sisquoc-points.txt)
* GPS readings from a [bike ride](data/ggmap_data/bike-ride.csv)
* A dataset of fish sampling [locations](data/ggmap_data/bc_sites.rds)

*Acknowledgement: The map-data in this course are obtained from Eric. C. Anderson's [GitHub page](https://github.com/eriqande)*

**End of document.**
