---
title: "Exploratory and differential expression analyses"
subtitle: "Workshop on RNA-Seq"
author: "`r paste0('<b>Ashfaq Ali and Markus Ringn√©r</b> | ',format(Sys.time(), '%d-%b-%Y'))`"

institute: NBIS, SciLifeLab
keywords: bioinformatics, course, scilifelab, nbis
output:
  xaringan::moon_reader:
    encoding: 'UTF-8'
    self_contained: false
    chakra: 'assets/remark-latest.min.js'
    css: 'assets/presentation.css'
    lib_dir: libs
    nature:
      ratio: '4:3'
      highlightLanguage: r
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
      slideNumberFormat: "%current%/%total%"
---
exclude: true
count: false

```{r,echo=FALSE,child="assets/header-presentation.Rmd"}
```

<!-- ------------ Only edit title, subtitle & author above this ------------ -->


```{r,include=FALSE}
# load the packages you need
library(dplyr)
library(tidyr)
#library(stringr)
library(ggplot2)
#library(plotly)
library(pheatmap)
library(DESeq2)
library(edgeR)
```

---
name: content
class: spaced

## Contents

* [Exploratory](#eda-cor)
* [DGE](#dge)
* [Vis](#vis)


---
```{r,include=FALSE}
cr <- read.table("./data/count_raw.txt",header=TRUE)
mr <- read.csv2("./data/metadata_raw.csv",header=TRUE,stringsAsFactors=F)
rownames(mr) <- mr$Sample_ID


cf <- cr[rowSums(edgeR::cpm(cr)>1) >= 2, ]
tpm <- function(counts,len) {
  x <- counts/len
  return(t(t(x)*1e6/colSums(x)))
}

#tpm
g <- read.delim("./data/human_genes.txt",header=T,stringsAsFactors=F)
g <- g[!duplicated(g$ensembl_gene_id),]
g$len <- g$end_position-g$start_position
rownames(g) <- g$ensembl_gene_id
igenes <- intersect(rownames(cf),g$ensembl_gene_id)
g1 <- g[igenes,]
cf1 <- cf[igenes,]
ct <- tpm(cf1,g1$len)

# deseq2
mr$Time <- factor(mr$Time)
d <- DESeqDataSetFromMatrix(countData=cf,colData=mr,design=~Time)
d <- DESeq2::estimateSizeFactors(d,type="ratio")
cd <- counts(d,normalized=TRUE)

# vst
mr$Time <- factor(mr$Time)
d <- DESeqDataSetFromMatrix(countData=cf,colData=mr,design=~Time)
d <- DESeq2::estimateSizeFactors(d,type="ratio")
d <- DESeq2::estimateDispersions(d)
cv <- as.data.frame(assay(varianceStabilizingTransformation(d,blind=T)),check.names=F)
```

name: eda-cor
## Exploratory Analyses



 - **Exploratory analyses on samples**
  - Detect outliers using PCA
  - Filter out low quality samples if possible
  - Map meta data on PCA plot to find any confounding factors
  - Sample to sample correlations

 - **Exploratory analyses on gene level estimates**

  - Gene counts/library size
  - Check if the treatment groups relate to library size
  - Gene dispersion estimate and visulizations

 - **Exploratory analyses on statitics**
  - MA plot of base mean vs. LFC
  - Histogram of p-values
  - Plot counts of individual genes

-
**Look for warning signs and take action on all levels and use visulizations as much as possible.**


---


## Exploratory | Correlation

* Correlation between samples

```{r,ech=FALSE,fig.height=4.5,fig.width=5.5}
dmat <- as.matrix(cor(cv,method="spearman"))
pheatmap::pheatmap(dmat,border_color=NA,annotation_col=mr[,"Time",drop=F],
         annotation_row=mr[,"Time",drop=F],annotation_legend=T)
```

---
name: eda-dist

## Exploratory | Distance

* Similarity between samples

```{r,echo=TRUE,fig.height=4.5,fig.width=5.5}
dmat <- as.matrix(dist(t(cv)))
pheatmap(dmat,border_color=NA,annotation_col=mr[,"Time",drop=F],
         annotation_row=mr[,"Time",drop=F],annotation_legend=T)
```

---
name: eda-pca

## Exploratory | PCA

* Relationship between samples

.pull-left-40[
```{r,echo=FALSE,fig.height=3.5,fig.width=3.5}
pcaobj <- prcomp(x=t(cv))
{par(mar=c(4,4,0.5,0))
barplot(round(pcaobj$sdev^2/sum(pcaobj$sdev^2)*100,2),las=2,
        names.arg=colnames(pcaobj$x),ylab="% Variance explained",
        xlab="PCA principal components")}
```
]

.pull-right-60[
```{r,echo=FALSE,fig.height=4,fig.width=4.5}
pcamat1 <- as.data.frame(pcaobj$x)
pcamat2 <- merge(pcamat1,mr,by=0)

ggplot(pcamat2,aes(PC1,PC2,colour=Time))+
  geom_point()+
  theme_bw()+
  theme(panel.border=element_blank(),
        axis.ticks=element_blank())
```

```{r,eval=FALSE,include=FALSE}
# plotly pca
library(plotly)
pcamat2$tt <- paste0("<b>ID: </b>",pcamat2$Sample_ID,"\n",
                   "<b>Name: </b>",pcamat2$Sample_Name)
p1 <- plot_ly(pcamat2,x=~PC1,y=~PC2,z=~PC3,text=~tt,hoverinfo="text")
add_markers(p1,color=~Time)
```

]

---
name: dge

## DGE

* Create the DESeq2 object

```{r}
library(DESeq2)
mr$Time <- factor(mr$Time)
d <- DESeqDataSetFromMatrix(countData=cf,colData=mr,design=~Time)
d
```

* Model must be factors

- `~var`
- `~covar+var`

---
name: dge-sf

## DGE | Size factors

* Normalisation factors are computed

```{r}
d <- DESeq2::estimateSizeFactors(d,type="ratio")
sizeFactors(d)
```

--
.pull-left-50[
```{r, echo=FALSE,fig.height=3.5,fig.width=3.5}
sizeFactors(d) %>% barplot(las =2, horiz = TRUE, axisnames = FALSE, main = "Size Factor")
```
]

.pull-right-50[
```{r, , echo=FALSE,fig.height=3.5,fig.width=3.5}
assays(d)$counts %>% colSums() %>% barplot(las=2, horiz = TRUE, axisnames = FALSE, main = "Total Counts")
```
]

---
name: dge-dispersion

## DGE | Dispersion

* We need a measure variability of gene counts

```{r}
dm <- apply(cd,1,mean)
dv <- apply(cd,1,var)
cva <- function(x) sd(x)/mean(x)
dc <- apply(cd,1,cva)
```

--

```{r,echo=FALSE,fig.height=2.5,fig.width=5}
h1 <- ggplot(data.frame(mean=log10(dm+1),var=log10(dv+1)),
       aes(mean,var))+
  geom_point(alpha=0.2)+
  geom_smooth(method="lm")+
  labs(x=expression('Log'[10]~'Mean counts'),y=expression('Log'[10]~'Variance'))+
  theme_bw()

h2 <- ggplot(data.frame(mean=log10(dm+1),var=dc),
       aes(mean,var))+
  geom_point(alpha=0.2)+
  geom_smooth(method="lm")+
  labs(x=expression('Log'[10]~'Mean counts'),y="Coefficient of variation")+
  theme_bw()
ggpubr::ggarrange(h1,h2,nrow=1,ncol=2)
```

* Dispersion is a measure of variability in gene expression for a given mean

---
name: dge-dispersion-2

## DGE | Dispersion

* D is unreliable for low mean counts
* Genes with similar mean values must have similar dispersion
* Estimate likely (ML) dispersion for each gene based on counts
* Fit a curve through the gene-wise estimates
* Shrink dispersion towards the curve

```{r,fig.height=3,fig.width=3}
d <- DESeq2::estimateDispersions(d)
{par(mar=c(4,4,1,1))
plotDispEsts(d)}
```

---
name: dge-test

## DGE | Test

* Log2 fold changes changes are computed after GLM fitting

```{r}
dg <- nbinomWaldTest(d)
resultsNames(dg)
```

--

* Use `results()` to customise/return results
  * Set coefficients using `contrast` or `name`
  * Filtering by fold change using `lfcThreshold`
  * `cooksCutoff` removes outliers
  * `independentFiltering`
  * `pAdjustMethod`
  * `alpha`

---
name: dge-test-2

## DGE | Test

```{r}
res1 <- results(dg,name="Time_t2_vs_t0",alpha=0.05)
summary(res1)
```

---
name: dge-test-3

## DGE | Test

```{r}
head(res1)
```

--

* Use `lfcShrink()` to correct fold changes for high dispersion genes

---
names: dge-lfcSrink-1

## Vis | LFC Shrink


```{r}
# Perform LFC shrinkage.

dg2<- lfcShrink(dg, res = res1, coef ="Time_t2_vs_t0", quiet = TRUE)
res2 <- dg2[which(dg2$padj<0.05 & abs(dg2$log2FoldChange) >0.58), ]
```
--
.pull-left-50[
```{r}
summary(res2)
```
]

.pull-right-50[
```{r}
res11 <- res1[which(res1$padj<0.05 & abs(res1$log2FoldChange) >0.58), ]
# Order based on original LFC
summary(res11)
```

]
---
names: dge-lfcSrink-2

## Vis | LFC Shrink

```{r}
# Order based on shrinked LFC
res2[order(abs(res2$log2FoldChange), decreasing = TRUE),c(1:2)][1:3,]
```





```{r}
# Order based on original LFC
res11[order(abs(res11$log2FoldChange),decreasing =
             TRUE),c(1:2)][1:3,]
```


***The ranks of top DE genes changes after `lfcSrink`***, note the **base means**.

---

name: vis-MAPlot

## Vis | PlotMA

***Plotting fold changes against mean counts can tell you about batch effects***


.pull-left-50[
```{r}
# Plot non shrinked estimates
DESeq2::plotMA(res1)
```
]
.pull-right-50[
```{r}
# Plot shrinked  LFC estimates
DESeq2::plotMA(dg2)
```
]

---

name: vis-volcano

## Vis | VolcanoPlot

```{r,fig.height=3.5,fig.width=3.5,block.title=TRUE}
ggplot()+
  geom_point(data=as.data.frame(res1),aes(x=log2FoldChange,y=-log10(padj)),col="grey80",alpha=0.5)+
  geom_point(data=filter(as.data.frame(res1),padj<0.05),aes(x=log2FoldChange,y=-log10(padj)),col="red",alpha=0.7)+
  geom_hline(aes(yintercept=-log10(0.05)),alpha=0.5)+
  theme_bw()
```

---
 
name: vis

## Vis | PlotCounts

.pull-left-50[

```{r,fig.height=5,fig.width=5,block.title=TRUE}
plotCounts(d,gene=rownames(res1)[1],intgroup="Time",normalized=F)

```

]

.pull-right-50[

```{r,fig.height=5,fig.width=5,block.title=TRUE}

plotCounts(d,gene=rownames(res1)[1],intgroup="Time",normalized=T)
```

]

---

name: vis-additional

## Vis | Additional

- Circose
- Networks (correlations)
- Intgerative Visualizations 

**You can take our [integrative omics](https://nbisweden.github.io/workshop_omics_integration/) **, ***April 20-20 in Lund***.

names: vis-counts

## Vis | Counts

***Plot individual gene counts***

.pull-left-50[
```{r,fig.height=5,fig.width=5,block.title=TRUE}
plotCounts(d,gene=rownames(res1)[1],intgroup="Time",normalized=F)
```
]

.pull-right-50[
```{r,fig.height=5,fig.width=5,block.title=TRUE}
plotCounts(d,gene=rownames(res1)[1],intgroup="Time",normalized=T)
```

]
<!-- --------------------- Do not edit this and below --------------------- -->

---

# Acknowledgements

* Slides Originally created by [Roy Francis](https://nbis.se/about/staff/roy-francis/) at NBIS
* [Normalising RNA-seq data](https://www.ebi.ac.uk/sites/ebi.ac.uk/files/content.ebi.ac.uk/materials/2012/121029_HTS/ernest_turro_normalising_rna-seq_data.pdf) by Ernest Turro
* RNA-seq analysis [Bioconductor vignette](http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html)
* [DGE Workshop](https://github.com/hbctraining/DGE_workshop/tree/master/lessons) by HBC training

---
name: end_slide
class: end-slide, middle
count: false

# Thank you. Questions?

```{r,echo=FALSE,child="assets/footer-presentation.Rmd"}
```

```{r,include=FALSE,eval=FALSE}
# manually run this to render this document to HTML
rmarkdown::render("presentation_exploratory_analyses.Rmd")
# manually run this to convert HTML to PDF
pagedown::chrome_print("presentation_exploratory_analyses.html",output="presentation_exploratory_analyses.pdf")
```
