<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
  <head>
    <title>Exploratory and differential expression analyses</title>
    <meta charset="utf-8" />
    <meta name="author" content="Ashfaq Ali and Markus Ringnér | 27-Nov-2019" />
    <meta name="keywords" content="bioinformatics, course, scilifelab, nbis" />
    <link href="libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
    <link href="libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
    <link id="font-awesome-1-attachment" rel="attachment" href="libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>
    <link rel="stylesheet" href="assets/presentation.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Exploratory and differential expression analyses
## Workshop on RNA-Seq
### <b>Ashfaq Ali and Markus Ringnér</b> | 27-Nov-2019
### NBIS, SciLifeLab

---

exclude: true
count: false


&lt;link href="https://fonts.googleapis.com/css?family=Roboto|Source+Sans+Pro:300,400,600|Ubuntu+Mono&amp;amp;subset=latin-ext" rel="stylesheet"&gt;
&lt;link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.3.1/css/all.css" integrity="sha384-mzrmE5qonljUremFsqc01SB46JvROS7bZs3IO2EmfFsd15uHvIt+Y8vEf7N7fWAU" crossorigin="anonymous"&gt;





&lt;!-- ------------ Only edit title, subtitle &amp; author above this ------------ --&gt;




---
name: content
class: spaced

## Contents

* [Exploratory](#eda-cor)
* [DGE](#dge)
* [Vis](#vis)


---


name: eda-cor
## Exploratory Analyses



 - **Exploratory analyses on samples**
  - Detect outliers using PCA
  - Filter out low quality samples if possible
  - Map meta data on PCA plot to find any confounding factors
  - Sample to sample correlations

 - **Exploratory analyses on gene level estimates**

  - Gene counts/library size
  - Check if the treatment groups relate to library size
  - Gene dispersion estimate and visulizations

 - **Exploratory analyses on statitics**
  - MA plot of base mean vs. LFC
  - Histogram of p-values
  - Plot counts of individual genes

-
**Look for warning signs and take action on all levels and use visulizations as much as possible.**


---


## Exploratory | Correlation

* Correlation between samples


```r
dmat &lt;- as.matrix(cor(cv,method="spearman"))
pheatmap::pheatmap(dmat,border_color=NA,annotation_col=mr[,"Time",drop=F],
         annotation_row=mr[,"Time",drop=F],annotation_legend=T)
```

&lt;img src="presentation_exploratory_analyses_files/figure-html/unnamed-chunk-4-1.svg" style="display: block; margin: auto auto auto 0;" /&gt;

---
name: eda-dist

## Exploratory | Distance

* Similarity between samples


```r
dmat &lt;- as.matrix(dist(t(cv)))
pheatmap(dmat,border_color=NA,annotation_col=mr[,"Time",drop=F],
         annotation_row=mr[,"Time",drop=F],annotation_legend=T)
```

&lt;img src="presentation_exploratory_analyses_files/figure-html/unnamed-chunk-5-1.svg" style="display: block; margin: auto auto auto 0;" /&gt;

---
name: eda-pca

## Exploratory | PCA

* Relationship between samples

.pull-left-40[
&lt;img src="presentation_exploratory_analyses_files/figure-html/unnamed-chunk-6-1.svg" style="display: block; margin: auto auto auto 0;" /&gt;
]

.pull-right-60[
&lt;img src="presentation_exploratory_analyses_files/figure-html/unnamed-chunk-7-1.svg" style="display: block; margin: auto auto auto 0;" /&gt;



]

---
name: dge

## DGE

* Create the DESeq2 object


```r
library(DESeq2)
mr$Time &lt;- factor(mr$Time)
d &lt;- DESeqDataSetFromMatrix(countData=cf,colData=mr,design=~Time)
d
```

```
## class: DESeqDataSet 
## dim: 14578 12 
## metadata(1): version
## assays(1): counts
## rownames(14578): ENSG00000000003 ENSG00000000419 ... ENSG00000266865
##   ENSG00000266876
## rowData names(0):
## colnames(12): Sample_1 Sample_2 ... Sample_11 Sample_12
## colData names(5): Sample_ID Sample_Name Time Replicate Cell
```

* Model must be factors

- `~var`
- `~covar+var`

---
name: dge-sf

## DGE | Size factors

* Normalisation factors are computed


```r
d &lt;- DESeq2::estimateSizeFactors(d,type="ratio")
sizeFactors(d)
```

```
##  Sample_1  Sample_2  Sample_3  Sample_4  Sample_5  Sample_6  Sample_7  Sample_8 
## 0.9003753 0.8437393 0.5106445 1.1276451 1.0941383 0.8133849 0.7553903 1.1744008 
##  Sample_9 Sample_10 Sample_11 Sample_12 
## 1.0189325 1.3642797 1.2325485 1.8555904
```

--
.pull-left-50[
&lt;img src="presentation_exploratory_analyses_files/figure-html/unnamed-chunk-11-1.svg" style="display: block; margin: auto auto auto 0;" /&gt;
]

.pull-right-50[
&lt;img src="presentation_exploratory_analyses_files/figure-html/unnamed-chunk-12-1.svg" style="display: block; margin: auto auto auto 0;" /&gt;
]

---
name: dge-dispersion

## DGE | Dispersion

* We need a measure variability of gene counts


```r
dm &lt;- apply(cd,1,mean)
dv &lt;- apply(cd,1,var)
cva &lt;- function(x) sd(x)/mean(x)
dc &lt;- apply(cd,1,cva)
```

--

&lt;img src="presentation_exploratory_analyses_files/figure-html/unnamed-chunk-14-1.svg" style="display: block; margin: auto auto auto 0;" /&gt;

* Dispersion is a measure of variability in gene expression for a given mean

---
name: dge-dispersion-2

## DGE | Dispersion

* D is unreliable for low mean counts
* Genes with similar mean values must have similar dispersion
* Estimate likely (ML) dispersion for each gene based on counts
* Fit a curve through the gene-wise estimates
* Shrink dispersion towards the curve


```r
d &lt;- DESeq2::estimateDispersions(d)
{par(mar=c(4,4,1,1))
plotDispEsts(d)}
```

&lt;img src="presentation_exploratory_analyses_files/figure-html/unnamed-chunk-15-1.svg" style="display: block; margin: auto auto auto 0;" /&gt;

---
name: dge-test

## DGE | Test

* Log2 fold changes changes are computed after GLM fitting


```r
dg &lt;- nbinomWaldTest(d)
resultsNames(dg)
```

```
## [1] "Intercept"      "Time_t2_vs_t0"  "Time_t24_vs_t0" "Time_t6_vs_t0"
```

--

* Use `results()` to customise/return results
  * Set coefficients using `contrast` or `name`
  * Filtering by fold change using `lfcThreshold`
  * `cooksCutoff` removes outliers
  * `independentFiltering`
  * `pAdjustMethod`
  * `alpha`

---
name: dge-test-2

## DGE | Test


```r
res1 &lt;- results(dg,name="Time_t2_vs_t0",alpha=0.05)
summary(res1)
```

```
## 
## out of 14578 with nonzero total read count
## adjusted p-value &lt; 0.05
## LFC &gt; 0 (up)       : 413, 2.8%
## LFC &lt; 0 (down)     : 696, 4.8%
## outliers [1]       : 0, 0%
## low counts [2]     : 2261, 16%
## (mean count &lt; 26)
## [1] see 'cooksCutoff' argument of ?results
## [2] see 'independentFiltering' argument of ?results
```

---
name: dge-test-3

## DGE | Test


```r
head(res1)
```

```
## log2 fold change (MLE): Time t2 vs t0 
## Wald test p-value: Time t2 vs t0 
## DataFrame with 6 rows and 6 columns
##                         baseMean     log2FoldChange              lfcSE
##                        &lt;numeric&gt;          &lt;numeric&gt;          &lt;numeric&gt;
## ENSG00000000003 490.017213130775  0.220619809383663   0.11276108378839
## ENSG00000000419 817.780657499214 0.0592719659601432  0.101481284662652
## ENSG00000000457  82.078767893562   0.20774862317291     0.220404896514
## ENSG00000000460  356.07160020304 -0.129186381676066  0.115139182335096
## ENSG00000001036 919.606750211436 0.0288826917736332 0.0851500691928605
## ENSG00000001084 529.593965301543  0.211964772147198 0.0929810563117207
##                              stat             pvalue              padj
##                         &lt;numeric&gt;          &lt;numeric&gt;         &lt;numeric&gt;
## ENSG00000000003  1.95652437854963 0.0504034135093186 0.263505451695364
## ENSG00000000419 0.584067950629293  0.559174596319362 0.830262316260442
## ENSG00000000457 0.942577168015478  0.345897222835914 0.689945926100397
## ENSG00000000460 -1.12200190288035  0.261861630829691 0.612624613845986
## ENSG00000001036 0.339197513841303  0.734460942036312 0.909638554355078
## ENSG00000001084   2.2796554540807 0.0226281312376009 0.159263252830589
```

--

* Use `lfcShrink()` to correct fold changes for high dispersion genes

---
names: dge-lfcSrink-1

## Vis | LFC Shrink



```r
# Perform LFC shrinkage.

dg2&lt;- lfcShrink(dg, res = res1, coef ="Time_t2_vs_t0", quiet = TRUE)
res2 &lt;- dg2[which(dg2$padj&lt;0.05 &amp; abs(dg2$log2FoldChange) &gt;0.58), ]
```
--
.pull-left-50[

```r
summary(res2)
```

```
## 
## out of 126 with nonzero total read count
## adjusted p-value &lt; 0.05
## LFC &gt; 0 (up)       : 46, 37%
## LFC &lt; 0 (down)     : 80, 63%
## outliers [1]       : 0, 0%
## low counts [2]     : 0, 0%
## (mean count &lt; 26)
## [1] see 'cooksCutoff' argument of ?results
## [2] see 'independentFiltering' argument of ?results
```
]

.pull-right-50[

```r
res11 &lt;- res1[which(res1$padj&lt;0.05 &amp; abs(res1$log2FoldChange) &gt;0.58), ]
# Order based on original LFC
summary(res11)
```

```
## 
## out of 382 with nonzero total read count
## adjusted p-value &lt; 0.05
## LFC &gt; 0 (up)       : 177, 46%
## LFC &lt; 0 (down)     : 205, 54%
## outliers [1]       : 0, 0%
## low counts [2]     : 0, 0%
## (mean count &lt; 26)
## [1] see 'cooksCutoff' argument of ?results
## [2] see 'independentFiltering' argument of ?results
```

]
---
names: dge-lfcSrink-2

## Vis | LFC Shrink


```r
# Order based on shrinked LFC
res2[order(abs(res2$log2FoldChange), decreasing = TRUE),c(1:2)][1:3,]
```

```
## log2 fold change (MAP): Time t2 vs t0 
##  
## DataFrame with 3 rows and 2 columns
##                         baseMean    log2FoldChange
##                        &lt;numeric&gt;         &lt;numeric&gt;
## ENSG00000120875 712.536673053803 -2.30638870244127
## ENSG00000142627 1572.22058641611 -1.98604476739038
## ENSG00000139289 1300.91800509536  -1.7345465752601
```






```r
# Order based on original LFC
res11[order(abs(res11$log2FoldChange),decreasing =
             TRUE),c(1:2)][1:3,]
```

```
## log2 fold change (MLE): Time t2 vs t0 
##  
## DataFrame with 3 rows and 2 columns
##                         baseMean    log2FoldChange
##                        &lt;numeric&gt;         &lt;numeric&gt;
## ENSG00000120875 712.536673053803 -2.76414118057783
## ENSG00000142627 1572.22058641611  -2.3878061230979
## ENSG00000139318 101.530389195382 -2.38675628664008
```


***The ranks of top DE genes changes after `lfcSrink`***, note the **base means**.

---

name: vis-MAPlot

## Vis | PlotMA

***Plotting fold changes against mean counts can tell you about batch effects***


.pull-left-50[

```r
# Plot non shrinked estimates
DESeq2::plotMA(res1)
```

&lt;img src="presentation_exploratory_analyses_files/figure-html/unnamed-chunk-24-1.svg" style="display: block; margin: auto auto auto 0;" /&gt;
]
.pull-right-50[

```r
# Plot shrinked  LFC estimates
DESeq2::plotMA(dg2)
```

&lt;img src="presentation_exploratory_analyses_files/figure-html/unnamed-chunk-25-1.svg" style="display: block; margin: auto auto auto 0;" /&gt;
]

---

name: vis-volcano

## Vis | VolcanoPlot


```r
ggplot()+
  geom_point(data=as.data.frame(res1),aes(x=log2FoldChange,y=-log10(padj)),col="grey80",alpha=0.5)+
  geom_point(data=filter(as.data.frame(res1),padj&lt;0.05),aes(x=log2FoldChange,y=-log10(padj)),col="red",alpha=0.7)+
  geom_hline(aes(yintercept=-log10(0.05)),alpha=0.5)+
  theme_bw()
```

&lt;img src="presentation_exploratory_analyses_files/figure-html/unnamed-chunk-26-1.svg" style="display: block; margin: auto auto auto 0;" /&gt;

---
 
name: vis

## Vis | PlotCounts

.pull-left-50[


```r
plotCounts(d,gene=rownames(res1)[1],intgroup="Time",normalized=F)
```

&lt;img src="presentation_exploratory_analyses_files/figure-html/unnamed-chunk-27-1.svg" style="display: block; margin: auto auto auto 0;" /&gt;

]

.pull-right-50[


```r
plotCounts(d,gene=rownames(res1)[1],intgroup="Time",normalized=T)
```

&lt;img src="presentation_exploratory_analyses_files/figure-html/unnamed-chunk-28-1.svg" style="display: block; margin: auto auto auto 0;" /&gt;

]

---

name: vis-additional

## Vis | Additional

- Circose
- Networks (correlations)
- Intgerative Visualizations 

**You can take our [integrative omics](https://nbisweden.github.io/workshop_omics_integration/) **, ***April 20-20 in Lund***.

names: vis-counts

## Vis | Counts

***Plot individual gene counts***

.pull-left-50[

```r
plotCounts(d,gene=rownames(res1)[1],intgroup="Time",normalized=F)
```

&lt;img src="presentation_exploratory_analyses_files/figure-html/unnamed-chunk-29-1.svg" style="display: block; margin: auto auto auto 0;" /&gt;
]

.pull-right-50[

```r
plotCounts(d,gene=rownames(res1)[1],intgroup="Time",normalized=T)
```

&lt;img src="presentation_exploratory_analyses_files/figure-html/unnamed-chunk-30-1.svg" style="display: block; margin: auto auto auto 0;" /&gt;

]
&lt;!-- --------------------- Do not edit this and below --------------------- --&gt;

---

# Acknowledgements

* Slides Originally created by [Roy Francis](https://nbis.se/about/staff/roy-francis/) at NBIS
* [Normalising RNA-seq data](https://www.ebi.ac.uk/sites/ebi.ac.uk/files/content.ebi.ac.uk/materials/2012/121029_HTS/ernest_turro_normalising_rna-seq_data.pdf) by Ernest Turro
* RNA-seq analysis [Bioconductor vignette](http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html)
* [DGE Workshop](https://github.com/hbctraining/DGE_workshop/tree/master/lessons) by HBC training

---
name: end_slide
class: end-slide, middle
count: false

# Thank you. Questions?



.end-text[
&lt;p&gt;R version 3.6.1 (2019-07-05)&lt;br&gt;&lt;p&gt;Platform: x86_64-pc-linux-gnu (64-bit)&lt;/p&gt;&lt;p&gt;OS: Ubuntu 19.10&lt;/p&gt;&lt;br&gt;

&lt;hr&gt;

&lt;span class="small"&gt;Built on : &lt;i class='fa fa-calendar' aria-hidden='true'&gt;&lt;/i&gt; 27-Nov-2019 at &lt;i class='fa fa-clock-o' aria-hidden='true'&gt;&lt;/i&gt; 03:07:55 PM&lt;/span&gt;  

&lt;b&gt;2019&lt;/b&gt; • [SciLifeLab](https://www.scilifelab.se/) • [NBIS](https://nbis.se/)
]
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="assets/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "4:3",
"highlightLanguage": "r",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"slideNumberFormat": "%current%/%total%"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
